### ISSUE TITLE
Compiler diagnostic suggests wrong keyword order

### ISSUE THREAD

**Issue by PoignardAzur** [rust-lang/rust#94879]
If I write this code:

```
#[cfg(FALSE)]
extern const fn hello() {}
```

[The compiler tells me:](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e60fe18af7d9f28837a7ad8c6201e80f)

```
error: expected `fn`, found keyword `const`
 --> src/lib.rs:2:8
  | ...
  = note: keyword order for functions declaration is `default`, `pub`, `const`, `async`, `unsafe`, `extern`
```

That's a great message, very helpful... except the order is wrong. If I write:

```
#[cfg(FALSE)]
default pub const async unsafe extern fn hello() {}
```

[The compiler gives me a weird error message.](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=19b5be0b9dea856e8cbbabea439f995e)

The correct order is `pub`, `default`, `const`, `async`, `unsafe`, `extern`; in other words, `pub` before everything else.

Indeed, the following code compiles without complaint:

```
#[cfg(FALSE)]
pub default const async unsafe extern fn hello() {}
```

(as well as other variants with `pub(crate)`, `crate`, etc)

The diagnostic suggestion should be fixed.

**Comment by PoignardAzur** [comment]
Side note: it's interesting that the error suggestion is triggered in some cases but not in others. Maybe the diagnostics should be standardized over the codebase.

From quick experimentation:

- If any of `const` `async` `unsafe` `extern` are in the wrong order, rustc gives the above message.
- If `pub` is in the wrong position, eg `const async pub unsafe` rustc gives another message:
    ```
    visibility `pub` must come before `const  async`: `pub const  async`
    ```
- If `default` is in the wrong position, rustc gives an unhelpful error message:
    ```
    error: expected `fn`, found `default`
    ```
    or
    ```
    error: `default` is not followed by an item
    ```

Might be worth a separate issue.
